### PR TITLE
[Distributed] allow get only distributed computed properties

### PR DESCRIPTION
**Pull request by ktoso**
The SE-0366 proposal said we'll support properties so start doing so by allowing to parse them.

The commit in the middle removes a Foundation dependency that's why the ad hoc Codable impl.

### CODE REVIEW

File: include/swift/AST/DiagnosticsSema.def [hunk 1]
```diff
@@ -4493,8 +4493,8 @@ ERROR(actor_isolated_non_self_reference,none,
         "from the main actor|from a non-isolated context}3",
         (DescriptiveDeclKind, DeclName, unsigned, unsigned, Type))
 ERROR(distributed_actor_isolated_non_self_reference,none,
-      "distributed actor-isolated %0 %1 can only be referenced inside the "
-      "distributed actor",
+      "distributed actor-isolated %0 %1 can not be accessed from a "
+      "non-isolated context",
```
Comment (ktoso) [inline include/swift/AST/DiagnosticsSema.def:4497]: These try to be more consistent with the rest of the current concurrency messages

File: include/swift/AST/DiagnosticsSema.def [hunk 2]
```diff
@@ -4566,11 +4566,11 @@ ERROR(distributed_actor_isolated_method,none,
...
 ERROR(distributed_actor_func_param_not_codable,none,
-      "parameter '%0' of type %1 in %2 does not conform to '%3'",
+      "parameter '%0' of type %1 in %2 does not conform to serialization requirement '%3'",
       (StringRef, Type, DescriptiveDeclKind, StringRef))
-ERROR(distributed_actor_func_result_not_codable,none,
-      "result type %0 of %1 does not conform to '%2'",
-      (Type, DescriptiveDeclKind, StringRef))
+ERROR(distributed_actor_target_result_not_codable,none,
+      "result type %0 of %1 %2 does not conform to serialization requirement '%3'",
+      (Type, DescriptiveDeclKind, Identifier, StringRef))
```
Comment (ktoso) [inline include/swift/AST/DiagnosticsSema.def:4573]: happy with the improvement here -- it is more clear WHY that conformance was expected

File: include/swift/AST/DiagnosticsSema.def [hunk 3]
```diff
@@ -4761,9 +4761,19 @@ ERROR(distributed_actor_user_defined_special_property,none,
...
+      (DeclName))
+ERROR(distributed_property_can_only_be_computed,none,
+      "static property %0 cannot be 'distributed', because it is not a computed "
+      "get-only property",
+      (DeclName))
 NOTE(distributed_actor_isolated_property,none,
-      "distributed actor state is only available within the actor instance", // TODO: reword in terms of isolation
-      ())
+      "access to this property is only permitted within the distributed actor %0",
+      (DeclName))
```
Comment (ktoso) [inline include/swift/AST/DiagnosticsSema.def:4776]: these again try to be more in line with how rest of swift concurrency messages such things

File: lib/Sema/TypeCheckDistributed.cpp [hunk 4]
```diff
@@ -214,36 +214,119 @@ bool swift::checkDistributedActorSystemAdHocProtocolRequirements(
...
+
+  if (resultType->isVoid())
+    return false;
+
+  // If the serialization requirement is specifically `Codable`
+  // we can issue slightly better warnings
+  llvm::SmallPtrSet<ProtocolDecl *, 2> codableRequirements = {
+      C.getProtocol(swift::KnownProtocolKind::Encodable),
+      C.getProtocol(swift::KnownProtocolKind::Decodable),
+  };
```
Comment (ktoso) [inline lib/Sema/TypeCheckDistributed.cpp:241]: TODO: make some util for this, this is done in more places

File: include/swift/AST/DiagnosticsSema.def (hunk 1 above)
Comment (xedin) [reply include/swift/AST/DiagnosticsSema.def:4497]: Nit: `cannot` and please drop `a` before `non-isolated context"

Comment (xedin) [reply include/swift/AST/DiagnosticsSema.def:4497]: Or `cannot access distributed actor-isolated %0 %1 from non-isolated context` even which is more inline with other diagnostics we have in Sema.

File: lib/Sema/TypeCheckDistributed.cpp [hunk 5]
```diff
@@ -214,36 +214,119 @@ bool swift::checkDistributedActorSystemAdHocProtocolRequirements(
...
+        getDistributedSerializationRequirementProtocols(actor);
+  } // TODO(distributed): need to handle ProtocolDecl too?
+
+  // If the requirement is exactly `Codable` we diagnose it ia bit nicer.
+  auto serializationRequirementIsCodable = false;
+  if (serializationRequirements.size() == 2) {
+      auto encodableType = C.getProtocol(KnownProtocolKind::Encodable);
+      auto decodableType = C.getProtocol(KnownProtocolKind::Decodable);
+      llvm::SmallPtrSet<ProtocolDecl *, 2> codableRequirements = {encodableType, decodableType};
+      serializationRequirementIsCodable =
```
Comment (xedin) [inline lib/Sema/TypeCheckDistributed.cpp:306]: Or simply - ` = serializationRequirements.count(encodableType) && serializationRequirements.count(decodableType)`?

File: include/swift/AST/DiagnosticsSema.def (hunk 1 above)
Comment (ktoso) [reply include/swift/AST/DiagnosticsSema.def:4497]: The message this tries to be same-as is:

```
ERROR(actor_isolated_non_self_reference,none,
        "actor-isolated %0 %1 can not be "
        "%select{referenced|mutated|used 'inout'}2 "
        "%select{on a non-isolated actor instance|"
        "from a Sendable function|from a Sendable closure|"
        "from an 'async let' initializer|from global actor %4|"
        "from the main actor|from a non-isolated context}3",
        (DescriptiveDeclKind, DeclName, unsigned, unsigned, Type))
```
So it is `actor-isolated %0 %1 can not be referenced from a non-isolated context`

File: include/swift/AST/DiagnosticsSema.def (hunk 3 above)
Comment (ktoso) [reply include/swift/AST/DiagnosticsSema.def:4776]: Sounds good! "this" is not useful, will do

File: include/swift/AST/DiagnosticsSema.def [hunk 6]
```diff
@@ -4761,9 +4761,19 @@ ERROR(distributed_actor_user_defined_special_property,none,
       "property %0 cannot be defined explicitly, as it conflicts with "
       "distributed actor synthesized stored property",
       (DeclName))
+ERROR(distributed_property_can_only_be_computed_get_only,none,
+      "'distributed' computed property %0 can only be have a 'get' implementation",
```
[the remark this answers is not in the document]
Comment (ktoso) [reply include/swift/AST/DiagnosticsSema.def:4765]: That sounds much better thank you!

File: include/swift/AST/DiagnosticsSema.def [hunk 7]
```diff
@@ -4761,9 +4761,19 @@ ERROR(distributed_actor_user_defined_special_property,none,
...
+      (DeclName))
+ERROR(distributed_property_cannot_be_static,none,
+      "'distributed' property %0 cannot be 'static'",
+      (DeclName))
+ERROR(distributed_property_can_only_be_computed,none,
+      "static property %0 cannot be 'distributed', because it is not a computed "
```
[the remark this answers is not in the document]
Comment (ktoso) [reply include/swift/AST/DiagnosticsSema.def:4771]: made this into  `      "static property %0 cannot be 'distributed', only computed properties can",`

File: lib/Sema/TypeCheckConcurrency.cpp [hunk 8]
```diff
@@ -1953,16 +1953,17 @@ namespace {
...
-      if (auto nominal = decl->getDeclContext()->getSelfNominalTypeDecl())
-        isDistributedActor = nominal->isDistributedActor();
+      auto nominal = decl->getDeclContext()->getSelfNominalTypeDecl();
+      bool isDistributedActor = false;
+      if (nominal) isDistributedActor = nominal->isDistributedActor();
 
       // FIXME: Make this diagnostic more sensitive to the isolation context of
       // the declaration.
       if (isDistributedActor) {
         if (dyn_cast<VarDecl>(decl)) {
```
[the remark this answers is not in the document]
Comment (ktoso) [reply lib/Sema/TypeCheckConcurrency.cpp:1963]: thx

File: lib/Sema/TypeCheckDistributed.cpp [hunk 9]
```diff
@@ -214,36 +214,119 @@ bool swift::checkDistributedActorSystemAdHocProtocolRequirements(
   return false;
 }
 
+static bool checkDistributedTargetResultType(
+    ModuleDecl *module,
+    ValueDecl *valueDecl,
+    llvm::SmallPtrSet<ProtocolDecl *, 2> serializationRequirements,
```
[the remark this answers is not in the document]
Comment (ktoso) [reply lib/Sema/TypeCheckDistributed.cpp:220]: right, thank you!

File: lib/Sema/TypeCheckDistributed.cpp (hunk 5 above)
Comment (ktoso) [reply lib/Sema/TypeCheckDistributed.cpp:306]: great, done :)

File: include/swift/AST/DiagnosticsSema.def (hunk 1 above)
Comment (ktoso) [reply include/swift/AST/DiagnosticsSema.def:4497]: Ok will adjust :) and eventually the concurrency ones

File: lib/Sema/TypeCheckDistributed.cpp [hunk 10]
```diff
@@ -214,36 +214,119 @@ bool swift::checkDistributedActorSystemAdHocProtocolRequirements(
...
+    bool diagnose) {
+  auto &C = valueDecl->getASTContext();
+
+  Type resultType;
+  if (auto func = dyn_cast<FuncDecl>(valueDecl)) {
+    resultType = func->mapTypeIntoContext(func->getResultInterfaceType());
```
[the remark this answers is not in the document]
Comment (ktoso) [reply lib/Sema/TypeCheckDistributed.cpp:226]: Seems to be necessary for this case:

```
/Users/ktoso/code/swift-project/swift/test/Distributed/distributed_actor_isolation.swift:83:20: error: unexpected error produced: result type 'T' of distributed instance method 'distReturnGeneric' does not conform to serialization requirement 'Codable'
  distributed func distReturnGeneric<T: Codable & Sendable>(item: T) async throws -> T { // ok
```

**Comment by ktoso** [comment]
@swift-ci please smoke test and merge

File: lib/Sema/TypeCheckDistributed.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (xedin) [reply lib/Sema/TypeCheckDistributed.cpp:225]: Hm, I guess it's just how conformsToProtocol works then, it needs the type as if it was used in the body of the function...

[the remark this answers is not in the document]
Comment (ktoso) [reply lib/Sema/TypeCheckDistributed.cpp:225]: Seems so :)

File: test/Distributed/distributed_actor_var_implicitly_async_throws.swift [hunk 11]
```diff
@@ -14,18 +14,18 @@ struct NotCodable {}
...
   var computed: String  {
     "normal"
   }
 
-  // expected-error@+1{{property 'dlet' cannot be 'distributed', because it is not a computed get-only property}}
+  // expected-error@+1{{property 'dlet' cannot be 'distributed', only computed properties can}}
   distributed let dlet: String = "illegal"
 
-  // expected-error@+1{{'distributed' computed property 'dletvar' can only be have a 'get' implementation}}
+  // expected-error@+1{{distributed property 'dletvar' cannot be 'distributed', only computed properties can}}
```
Comment (xedin) [inline test/Distributed/distributed_actor_var_implicitly_async_throws.swift:28]: Nit: This one is a bit tautological, should probably replace `distributed` with `stored` in this case so it reads - `stored property 'dletvar' cannot be 'distributed', only computed properties can`

File: test/Distributed/distributed_actor_var_implicitly_async_throws.swift [hunk 12]
```diff
@@ -14,18 +14,18 @@ struct NotCodable {}
 
 distributed actor D {
 
-  // expected-note@+1{{access to this property is only permitted within the distributed actor 'D'}}
+  // expected-note@+1{{access to property 'normal' is only permitted within distributed actor 'D'}}
   var normal: String = "normal"
 
-  // expected-note@+1{{access to this property is only permitted within the distributed actor 'D'}}
+  // expected-note@+1{{access to property 'computed' is only permitted within distributed actor 'D'}}
```
Comment (xedin) [inline test/Distributed/distributed_actor_var_implicitly_async_throws.swift:20]: `permitted only within` - i think that's the correct use of this modifier.

**Pull request merged**